ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs

```diff
@@ -558,25 +582,108 @@ peerSelectionGovernorLoop tracer
+
+        let !outboundConnectionsState =
+              case (promiscuousMode, bootstrapPeersFlag st'') of
+                {-
+                 -- genesis mode
+                (NonPromiscuousMode, _)
+                          |  numberOfActiveBigLedgerPeers >= targetNumberOfActiveBigLedgerPeers
+                          -> TrustedStateWithExternalPeers
+
+                          |  otherwise
+                          -> UntrustedState
+                -}
+
+                (NonPromiscuousMode, _)
+                  |  -- we are only connected to trusted local root
+                     -- peers or bootstrap peers
+                     viewEstablishedPeers `Set.isSubsetOf` (viewEstablishedBootstrapPeers <> trustableLocalRootSet)
+                  -> TrustedStateWithExternalPeers
+
+                  | otherwise
+                  -> UntrustedState
+
+                 -- bootstrap mode
+                (PromiscuousMode, UseBootstrapPeers {})
+                          |  -- we are only connected to trusted local root
+                             -- peers or bootstrap peers
+                             viewEstablishedPeers `Set.isSubsetOf` (viewEstablishedBootstrapPeers <> trustableLocalRootSet)
+                             -- there's at least one active bootstrap peer
+                          ,  not (Set.null viewActiveBootstrapPeers)
+                          -> TrustedStateWithExternalPeers
+
+                          |  otherwise
+                          -> UntrustedState
+
+                 -- praos mode with public roots
+                 -- TODO: we need to pattern match on `Praos` / `Genesis` mode,
+                 -- see below.
+                (PromiscuousMode, DontUseBootstrapPeers)
+                          | -- we are only connected to trusted local root
+                            -- peers and at least one public root peer
+                            --
+                            -- Since the node is promiscuous and don't use
+                            -- bootstrap peers, it must be using ledger peers.
+                            -- It's likely to have non-zero targets for big
+                            -- ledger peers or ledger peers, in which case the
+                            -- condition below likely will not be satisfied,
+                            -- and thus `UntrustedState` will be returned.
+                            let establishedPublicRootPeers =
+                                  viewEstablishedPeers
+                                  `Set.intersection`
+                                  Map.keysSet (getPublicConfigPeers (publicRootPeers st''))
+                          , viewEstablishedPeers `Set.isSubsetOf` (establishedPublicRootPeers <> trustableLocalRootSet)
+                          , not (null establishedPublicRootPeers)
+                          -> TrustedStateWithExternalPeers
+
+                          |  otherwise
+                          -> UntrustedState
```

Comment:
Can this logic be moved to `readPromiscuousMode` ?

Reply:
I'll move it to it's own function, `readAssociationMode` (how `readPromiscuousMode` was renamed) it's good as it stands (small, easy to understand and useful for testing).